apm-sniffer/apm-sdk-plugin/dubbo-3-conflict-patch/src/main/java/org/apache/skywalking/apm/plugin/asf/dubbo3/patch/WrapperInstrumentation.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.skywalking.apm.plugin.asf.dubbo3.patch;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.StaticMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassStaticMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * The dubbo conflict plugin resolver the problem about the wrapper class generated by Dubbo core cannot be compiled
+ * successfully. As we known, The wrapper class traverses all the methods. In usual, it works unless this class has been
+ * enhanced by Skywalking. The javasist cannot found the `EnhanceInstance` method when generate.
+ * <p>
+ * The plugin excludes {@link org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance} methods
+ * to ensure the correct compilation of the code.
+ */
+public class WrapperInstrumentation extends ClassStaticMethodsEnhancePluginDefine {
+    @Override
+    public StaticMethodsInterceptPoint[] getStaticMethodsInterceptPoints() {
+        return new StaticMethodsInterceptPoint[] {
+            new StaticMethodsInterceptPoint() {
+                @Override
+                public ElementMatcher<MethodDescription> getMethodsMatcher() {
+                    return named("makeWrapper");
+                }
+
+                @Override
+                public String getMethodsInterceptor() {
+                    return "org.apache.skywalking.apm.plugin.asf.dubbo3.patch.MakeWrapperInterceptor";
+                }
+
+                @Override
+                public boolean isOverrideArgs() {
+                    return false;
+                }
+            }
+        };
+    }
+
+    @Override
+    protected ClassMatch enhanceClass() {
+        return byName("org.apache.dubbo.common.bytecode.Wrapper");
```

Comment:
I can see this is as same as Dubbo v2's `WrapperInstrumentation`. Is this duplicated **conflict patch**?

I can't see `dubbo` `v2->v3` from namespace level in `MakeWrapperInterceptor`, and there is no witness class in this instrumentation. Should we just simple change `dubbo-2.7.x-conflict-patch` to `dubbo-2.7.x-3.x-conflict-patch`

Reply:
Yes, indeed there is no difference at namespace level between v2 and v3. Actually dubbo 3 even gave an announcement that no need to touch a single line of code just update version from 2.x to 3.x in dependency of pom.xml, and the migration to v3 is done.

Reply:
Then, could you share why dubbo 2.x plugin doesn't work on dubbo 3.x?

Reply:
It because, for classes and methods expose to user side (rpc level) is compatible with v2, but the internal dosen't.
For example we can't rely on `RpcContext.getContext()` with v3.

Reply:
OK, then. In theory(you could test and verify), if this patch classes(instrumentation + interceptor) are exactly 100% same as v2, then this is not required.

> for classes and methods expose to user side (rpc level) is compatible with v2, but the internal dosen't.
For example we can't rely on RpcContext.getContext() with v3.

This would make the plugin for Dubbo 3 is required, and you should add `witness class` to dubbo 3 and dubbo 2.7 plugins, because they share the same target, you should make the agent core knows when should activate this instrumentation, and pick the right interceptor in the runtime.

Do you knoow what is the `witness class`? If not, read this, https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/java-plugin-development-guide/#implement-plugin, search this `4. Set up witnessClasses and/or witnessMethods if the instrumentation has to be activated in specific versions.`

Reply:
Then follow [this doc](https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/plugin-test/) and take a reference from what I shared for dubbo 2.7 plugin test, you could add a new for dubbo 3.0.

Once both of them pass the test, we should be good.

Reply:
Ok, just need some time, won't be too lang.